flink-ml-lib/src/main/java/org/apache/flink/ml/feature/hashingtf/HashingTF.java

```diff
@@ -0,0 +1,195 @@
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        ResolvedSchema tableSchema = inputs[0].getResolvedSchema();
+
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(tableSchema);
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(
+                                inputTypeInfo.getFieldTypes(), SparseVectorTypeInfo.INSTANCE),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
+
+        DataStream<Row> output =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new HashTFFunction(getInputCol(), getBinary(), getNumFeatures()),
+                                outputTypeInfo);
+        return new Table[] {tEnv.fromDataStream(output)};
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    public static HashingTF load(StreamTableEnvironment tEnv, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /** The main logic of {@link HashingTF}, which converts the input to a sparse vector. */
+    public static class HashTFFunction implements MapFunction<Row, Row> {
+        private final String inputCol;
+        private final boolean binary;
+        private final int numFeatures;
+
+        public HashTFFunction(String inputCol, boolean binary, int numFeatures) {
+            this.inputCol = inputCol;
+            this.binary = binary;
+            this.numFeatures = numFeatures;
+        }
+
+        @Override
+        public Row map(Row row) throws Exception {
+            Object inputObj = row.getField(inputCol);
+
+            Iterable<Object> inputList;
+            if (inputObj.getClass().isArray()) {
+                inputList = Arrays.asList((Object[]) inputObj);
+            } else if (inputObj instanceof Iterable) {
+                inputList = (Iterable<Object>) inputObj;
+            } else {
+                throw new IllegalArgumentException(
+                        "The input format is not supported. "
+                                + "Supported options are Array and Iterable.");
+            }
+
+            Map<Integer, Integer> map = new HashMap<>();
+            for (Object obj : inputList) {
+                int hashValue = hash(obj);
+                int index = nonNegativeMod(hashValue, numFeatures);
+                if (map.containsKey(index)) {
+                    if (!binary) {
+                        map.put(index, map.get(index) + 1);
+                    }
+                } else {
+                    map.put(index, 1);
+                }
+            }
+
+            // Converts to map to a sparse vector.
```

Comment:
nit: "from map to a sparse vector".

Reply:
emmm, what's wrong with the current comment?